src/target/rp.c

```diff
@@ -554,56 +611,216 @@ static void rp_spi_read(
+// These can be the same buffer.
+// If tx is NULL then send zeroes.
+// If rx is NULL then all read data will be dropped.
+//
+// If rx_skip is nonzero, this many bytes will first be consumed from the FIFO,
+// before reading a further count bytes into *rx.
+// E.g. if you have written a command+address just before calling this function.
+static void rp_flash_put_get(target *t, const uint8_t *tx, uint8_t *rx, size_t count, size_t rx_skip) {
+    // Make sure there is never more data in flight than the depth of the RX
+    // FIFO. Otherwise, when we are interrupted for long periods, hardware
+    // will overflow the RX FIFO.
+    const uint max_in_flight = 16 - 2; // account for data internal to SSI
+    size_t tx_count = count;
+    size_t rx_count = count;
+    while (tx_count || rx_skip || rx_count) {
+        // NB order of reads, for pessimism rather than optimism
+        uint32_t tx_level = target_mem_read32(t, RP_SSI_TXFLR);
+        uint32_t rx_level = target_mem_read32(t, RP_SSI_RXFLR);
+        bool did_something = false; // Expect this to be folded into control flow, not register
+        if (tx_count && tx_level + rx_level < max_in_flight) {
+            target_mem_write32(t, RP_SSI_DR0, (uint32_t) (tx ? *tx++ : 0));
+            --tx_count;
+            did_something = true;
+        }
+        if (rx_level) {
+            uint8_t rxbyte = target_mem_read32(t, RP_SSI_DR0);
+            did_something = true;
+            if (rx_skip) {
+                --rx_skip;
+            } else {
+                if (rx)
+                    *rx++ = rxbyte;
+                --rx_count;
+            }
+        }
+        // APB load costs 4 cycles, so only do it on idle loops (our budget is 48 cyc/byte)
+        if (!did_something && rp_flash_was_aborted(t))
+            break;
+    }
+    rp_spi_chip_select(t, RP_GPIO_QSPI_CS_DRIVE_HIGH);
+}
+
+// Sequence:
+// 1. CSn = 1, IO = 4'h0 (via pulldown to avoid contention), x32 clocks
+// 2. CSn = 0, IO = 4'hf (via pullup to avoid contention), x32 clocks
+// 3. CSn = 1 (brief deassertion)
+// 4. CSn = 0, MOSI = 1'b1 driven, x16 clocks
+//
+// Part 4 is the sequence suggested in W25X10CL datasheet.
+// Parts 1 and 2 are to improve compatibility with Micron parts
+static void rp_flash_exit_xip(target *t)
+{
+    uint8_t buf[2];
+    buf[0] = 0xff;
+    buf[1] = 0xff;
+
+	rp_flash_init_spi(t);
+
+	uint32_t padctrl_save = target_mem_read32(t, RP_PADS_QSPI_GPIO_SD0);
+    uint32_t padctrl_tmp = (padctrl_save
+                            & ~(RP_PADS_QSPI_GPIO_SD0_OD_BITS | RP_PADS_QSPI_GPIO_SD0_PUE_BITS |
+                                RP_PADS_QSPI_GPIO_SD0_PDE_BITS)
+                           ) | RP_PADS_QSPI_GPIO_SD0_OD_BITS | RP_PADS_QSPI_GPIO_SD0_PDE_BITS;
+
+    // First two 32-clock sequences
+    // CSn is held high for the first 32 clocks, then asserted low for next 32
+    rp_spi_chip_select(t, RP_GPIO_QSPI_CS_DRIVE_HIGH);
+    for (int i = 0; i < 2; ++i) {
```

Comment:
Please use an unsigned type (size_t would be suitable) for this loop

Reply:
Fixed!